### ISSUE TITLE
Incorrect value of "finish_reason" when streaming

### ISSUE THREAD

**Issue by longseespace** [ollama/ollama#4357]
### What is the issue?

When streaming using OpenAI server, the "finish_reason" is an empty which is incorrect. It should be one of the values from OpenAI, or null.

```
data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" Hello"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" there"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"!"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" I"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"'"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"m"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" B"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"olt"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"AI"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":","},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" your"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" helpful"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" assistant"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"."},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" How"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" can"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" I"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" help"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" you"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" today"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"?"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" Let"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" me"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" know"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" if"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" you"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" have"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" any"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" questions"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427619,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" or"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" tasks"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" you"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" need"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" assistance"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" with"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"."},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" I"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"'"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"ll"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" do"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" my"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" best"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" to"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" provide"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" you"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" with"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" accurate"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" and"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" tim"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"ely"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" information"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"."},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" Go"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" ahead"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" and"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" give"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" me"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" a"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" try"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"!"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":" "},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427620,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"😊"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427621,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":"💡"},"finish_reason":""}]}

data: {"id":"chatcmpl-693","object":"chat.completion.chunk","created":1715427621,"model":"mistral:latest","system_fingerprint":"fp_ollama","choices":[{"index":0,"delta":{"role":"assistant","content":""},"finish_reason":"stop"}]}

data: [DONE]
```

I think this was the source of the issue though I might be wrong:

https://github.com/ollama/ollama/commit/cfa84b8470837ecb418d1668858fe06c35c01d34#diff-e09be2f542df58a8abd41a5061837090a6b630a5f16f491b0e7a61e9dc429679

### OS

macOS

### GPU

Apple

### CPU

Apple

### Ollama version

0.1.36

**Comment by ziggornif** [comment]
I think we can restore the valid response payload with something like that : 

```go
FinishReason: func(reason string) *string {
	if len(reason) > 0 {
		return &reason
	}
	return nil
}(r.DoneReason),
```

**Comment by jmorganca** [comment]
@ziggornif thanks for the suggestion!

**Comment by GusBot69** [comment]
This bug has been causing cascading issues across every downstream project that uses Ollama's OpenAI-compatible API. Specifically hitting this right now with GLM-5.2 on Ollama Cloud:

- **Hermes Agent** (#10711, #14572, #72316): had to build an entire `_is_ollama_glm_backend()` heuristic to detect truncated responses that Ollama reports as `finish_reason="stop"`. The heuristic keeps false-positive-ing on legitimate complete responses, causing infinite continuation loops where the agent produces duplicate messages. Three open PRs trying to work around it (#72332, #73223, #60996) and one already-merged fix that regressed (#53574).
- **LiteLLM** (#25825): same issue, `finish_reason` always `"stop"` when truncated by `max_tokens`/`num_predict`.
- **LangChain** (#37370): `finish_reason` can be `None` or a list like `["END"]` instead of the OpenAI spec values.

The fix is straightforward: when the response is cut off by a token limit (`num_predict`), return `finish_reason="length"` instead of `"stop"`. That single change would let every downstream project delete their workarounds and stop building heuristics to guess whether Ollama is telling the truth.

This has been open for months. Multiple projects are bleeding engineering hours building band-aids for something that is one conditional statement to fix upstream. Please prioritize this.
